File: daily/Sliding Window/1358-Number-of-substrings-containing-all-three-characters.go
package daily

func numberOfSubstrings(s string) int {
	n := len(s)
	left := 0
	res := 0
	count := map[byte]int{'a': 0, 'b': 0, 'c': 0}

	for right := 0; right < n; right++ {
		count[s[right]]++
		for count['a'] > 0 && count['b'] > 0 && count['c'] > 0 {
			res += n - right
			count[s[left]]--
			left++
		}
	}
	return res
}

File: daily/Sliding Window/1358-Number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        n = len(s)
        left = res = 0
        count = { 'a': 0, 'b': 0, 'c': 0 }
        for right in range(n):
            count[s[right]] += 1
            while count['a'] > 0 and count['b'] > 0 and count['c'] > 0:
                res += n - right
                count[s[left]] -= 1
                left += 1
        return res